### ISSUE TITLE
mishandles "mul nsw" semantics?

### ISSUE THREAD

**Issue by wujingyue** [nunoplopes/alive#28]
Optimization: 1
Precondition: true
  %1 = add nsw i32 %b, 3
  %2 = mul nsw i32 %1, %s
  %3 = add nsw i32 %b, 15
  %4 = mul nsw i32 %3, %s
=>
  %1 = add nsw i32 %b, 3
  %2 = mul nsw i32 %1, %s
  %5 = mul %s, 12
  %4 = add nsw i32 %2, %5

ERROR: Domain of poisoness of Target is smaller than Source's for i32 %4

Example:
%b i32 = 0xFFFFFFF0 (4294967280, -16)
%s i32 = 0xC020013A (3223322938, -1071644358)
%1 i32 = 0xFFFFFFF3 (4294967283, -13)
%2 i32 = 0x3E5FF00E (1046474766)
%3 i32 = 0xFFFFFFFF (4294967295, -1)
%5 i32 = 0x01800EB8 (25169592)
Source value: 0x3FDFFEC6 (1071644358)
Target value: poison

I ran into a seemingly bug. The transformation assumes %1 * %s doesn't sign-overflow, but In the counter example %1 * %s (-13 * -1071644358) sign-overflows.

**Comment by nunoplopes** [comment]
The issue is that there's no dependency between %4 and %2 on the source template, and therefore the poison constraint doesn't flow there.  Poison flows through the def-use chain only.
Do you agree with this?

**Comment by wujingyue** [comment]
I used to incorrectly think "a value becomes poison" is the end of the world and that the compiler can assume no values are poison. However, a poison value can still be operated on; it just typically propagates the poisoness to other values. 

With that, what alive does here makes sense. With the example alive gives, in the source, 

%1: non-poison
%2: poison
%3: non-poison
%4: non-poison

and in the target

%1: non-poison
%2: poison
%3: non-poison
%4: poison

Therefore, the transformation is incorrect because %4 is poison in the target but not the source. 

Is my understanding correct?

**Comment by wujingyue** [comment]
A more general question for you. One thing StraightLineStrengthReduce does is

X = (B + i) * S
Y = (B + j) * S // i and j are constants. 
=>
Y = X + (j - i) * S

However, according to my new theory, this transformation is in general wrong because it makes Y depends on X. If in the source X is poison and Y is not, then in the target both of them are poison! 

Any thoughts? Am I too pessimistic?

**Comment by nunoplopes** [comment]
Regarding your 1st comment, I believe your understanding is correct.  Since there's no poison flowing into %4 in the source, but there is in the target, then the transformation is incorrect.

Regarding the 2nd comment, it will depend on the nsw/nuw flags that the multiplication and addition have. I didn't try, but it seems to me that in the worst case some of the flags may have to be dropped. I think you can try a few examples to see which break and which do not  (I haven't read SLSR's code yet, so I dunno how aggressive it is in propagating nsw/nuw flags).
A second note is that David Majnemer is proposing sligthly different semantics for nsw/nuw, removing poison altogether, so some results may change slightly if the new semantics for LLVM IR are adopted.
The bottom line is that it looks like you need to be careful with handling nsw/nuw in X and in Y.

**Comment by wujingyue** [comment]
Even if SLSR doesn't propagate nsw at all, it might still do wrong. 

For example,
```
X = (B +nsw i) * S
Y = (B + j) * S
=>
X = (B +nsw i) * S
Y = X + (j - i) * S
```

Say B +nsw i triggers overflow and is thus poison. Therefore, in the source, X is poison and Y in the source is not (because none of its subexpressions involve nsw). In contrast, in the target, X is poison and Y is also poison because it depends on X. This issue could be fixed by stripping the nsw in X as well (i.e., X = (B + i) * S), but it's quite heavily-handed -- why optimizing Y makes X imprecise? 

In a high level, this issue seems to bite all such optimizations that introduce new dependencies. This sounds scary to me :)

Back up one more step, a more fundamental issue here is the definition of poisoness in LLVM is weaker than C/C++'s undefined behavior. C/C++'s undefined behavior immediately "free up" the compiler. However, a poison value in LLVM may be propagated further until it is observed. I found [some detailed discussion](https://groups.google.com/forum/#!topic/llvm-dev/sDYaYV_ZF-g) back in 2011 that decided the current semantics of nsw. 

Can you point me to David Majnemer's proposal? I didn't manage to find it. Thank you!

**Comment by nunoplopes** [comment]
That's correct.  See what Alive has to say:
```
  %X0 = add nsw %B, %i
  %X = mul %X0, %S
  %Y0 = add %B, %j
  %Y = mul %Y0, %S
=>
  %X0 = add nsw %B, %i
  %X = mul %X0, %S
  %t0 = sub %j, %i
  %t1 = mul %t0, %S
  %Y = add %X, %t1


ERROR: Domain of poisoness of Target is smaller than Source's for i4 %Y

Example:
%B i4 = 0x8 (8, -8)
%i i4 = 0x8 (8, -8)
%S i4 = 0x0 (0)
%j i4 = 0x0 (0)
%X0 i4 = 0x0 (0)
%X i4 = 0x0 (0)
%Y0 i4 = 0x8 (8, -8)
%t0 i4 = 0x8 (8, -8)
%t1 i4 = 0x0 (0)
Source value: 0x0 (0)
Target value: poison
```

Anyway, this is the reason why passes like GVN need to drop most nsw/nuw flags.

And yes, LLVM *has* to be weaker than C++'s standard for other reasons.  We will hopefully write up about that soon.
I couldn't readily find David's proposal, but this (long)  thread hints at it: http://lists.cs.uiuc.edu/pipermail/llvmdev/2015-January/081310.html
